### PR TITLE
add support for JSONAPISerializer

### PR DESCRIPTION
**Pull request by Azdaroth**
Requires some more testing and probably acceptance tests with different models to use both REST and JSONAPI serializers, but the new implementation works.

### CODE REVIEW

**Comment by Azdaroth** [comment]
@ZenCocoon What do you think? Should we add anything more in the docs? Does do the customization look ok?

File: README.md [hunk 1]
```diff
@@ -46,6 +49,97 @@ In your template you can do
...
+import parentSerializerMixin from 'ember-data-partial-model/mixins/serializer';
+const { Mixin } = Ember;
+
+export default Mixin.create(parentSerializerMixin, {
+  serializeBelongsTo: function(snapshot, json, relationship) {
+     // overwrite this function to handle serialization of
+     // the relationship for extension, most likely you want to
+     // return null for partial extension. Required.
+     //  if (relationship.options.isPartialExtension) {
+     //   return null;
```
Comment (ZenCocoon) [inline README.md]: small indentation fix needed

File: addon/mixins/jsonapi-serializer.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
+import Ember from 'ember';
+import parentSerializerMixin from 'ember-data-partial-model/mixins/serializer';
```
Comment (ZenCocoon) [inline addon/mixins/jsonapi-serializer.js]: how do you feel about:

``` js
import baseSerializerMixin from 'ember-data-partial-model/mixins/base_serializer'; // note the filename change
```

File: README.md [hunk 3]
```diff
@@ -46,6 +49,97 @@ In your template you can do
...
+
+  _normalizePartialRelationship: function(modelClass, hash, descriptor) {
+    // add this function to handle normalization of partial extension as a relationship. Required.
+  },
+
+  _copyAttributesFromPartialToParent: function(ourHash, partialData) {
+    // add this function to handle copying attributes from partial extension to top-level hash. Required.
+  },
+
+  _modifyHashBasedOnDescriptorKey: function(ourHash, descriptorKey) {
```
Comment (ZenCocoon) [inline README.md]: The base serializer is already defining this function so their's no need to define it unless you need to do something with it.

File: addon/mixins/jsonapi-serializer.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    hash.relationships = hash.relationships || {};
+    hash.relationships[descriptor.key] = {
+      data: {
+        type: descriptor.type,
+        id: context.extractId(modelClass, hash)
+      }
+    };
+  },
+
+  _copyAttributesFromPartialToParent: function(ourHash, partialData) {
```
Comment (ZenCocoon) [inline addon/mixins/jsonapi-serializer.js]: The naming of `ourHash` feels odd, what about `resourceHash`? but also `partialHash`

**Comment by Azdaroth** [comment]
@ZenCocoon Fixed.

File: README.md [hunk 5]
```diff
@@ -46,6 +49,96 @@ In your template you can do
...
+If you haven't customized any of your adapters, serializers or store, the addon will work out of the box. By default is uses JSONAPISerializer and JSONAPIAdapter.
+
+If you made any customization, you will need to include proper mixins. For adapter:
+
+```js
+// app/adapters/application.js
+
+import DS from 'ember-data';
+import PartialModelAdapter from 'ember-data-partial-model/mixins/adapter';
+const { RESTAdapter } = DS;
```
Comment (ZenCocoon) [inline README.md]: I would default this example to the `JSONAPIAdapter` as default in Ember Data now.

File: README.md [hunk 6]
```diff
@@ -46,6 +49,96 @@ In your template you can do
...
+```
+
+For serializer (both RESTSerializer and JSONAPISerializer are available):
+
+```js
+// app/adapters/serializer.js
+
+import DS from 'ember-data';
+import PartialModelRESTSerializer from 'ember-data-partial-model/mixins/rest-serializer';
+const { RESTSerializer } = DS;
```
Comment (ZenCocoon) [inline README.md]: Show the `JSONAPISerializer` setup first, that's the default one.

File: addon/mixins/serializer.js [hunk 7]
```diff
@@ -1,52 +1,31 @@
 import Ember from 'ember';
```
Comment (ZenCocoon) [inline addon/mixins/serializer.js]: What about calling this file `addon/mixins/base_serializer.js` instead of `addon/mixins/serializer.js`?

File: tests/acceptance/rentals-test.js [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+    assert.equal(find('.name').val(), "Rental 0");
+    assert.equal(find('.description').val(), "Awesome Villa-0");
+  });
+});
+
+test("I can create a new rental and it's partial model values", function(assert) {
+  visit('/rentals/new');
+
+  fillIn('.name', 'VillasThalassa');
+  fillIn('.description', 'Awesome villa');
```
Comment (ZenCocoon) [inline tests/acceptance/rentals-test.js]: To be picky, would be `villas` here :)

File: tests/acceptance/rentals-test.js [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+    assert.equal(currentRouteName(), 'rentals.rental');
+    assert.equal(find('.name').val(), "VillasThalassa");
+    assert.equal(find('.description').val(), "Awesome villa");
+  });
+});
+
+test("I can edit an existing rental and it's partial model values", function(assert) {
+  visit('/rentals/1');
+
+  fillIn('.name', 'VillasSivota');
```
Comment (ZenCocoon) [inline tests/acceptance/rentals-test.js]: To be picky, would be `Villa Sivota` here :)

File: tests/dummy/app/templates/rentals.hbs [hunk 10]
```diff
@@ -0,0 +1,10 @@
+<ul>
+  {{#each model as |rental|}}
+    <li>
+    {{#link-to 'rentals.rental' rental}}{{ rental.name }}{{/link-to}}
```
Comment (ZenCocoon) [inline tests/dummy/app/templates/rentals.hbs]: Would be best indeed with 2 more space. Same bellow

**Comment by Azdaroth** [comment]
@ZenCocoon fixed.

File: README.md [hunk 11]
```diff
@@ -46,6 +49,97 @@ In your template you can do
 {{user.twitter}}
 ```
 and it will populate once the full user is loaded
+
+## How to use
+If you haven't customized any of your adapters, serializers or store, the addon will work out of the box. By default is uses JSONAPISerializer and JSONAPIAdapter.
```
Comment (igorT) [inline README.md]: it*

File: README.md [hunk 12]
```diff
@@ -46,6 +49,97 @@ In your template you can do
...
+
+If you made any customization, you will need to include proper mixins. For adapter:
```
Comment (igorT) [inline README.md]: You could also keep having it work by default for RestAdapter people by looking up the application adapter at runtime

File: README.md [hunk 13]
```diff
@@ -46,6 +49,97 @@ In your template you can do
...
+
+export default JSONAPIAdapter.extend(PartialModelAdapter);
+
+```
+
+For serializer (both JSONAPISerializer and RESTSerializer are available):
+
+
+```js
+// app/adapters/serializer.js
```
Comment (igorT) [inline README.md]: This path isn't correct is it?

Comment (Azdaroth) [inline README.md]: Right, should be `application.js`, same thing for the serializer below.

File: addon/mixins/jsonapi-serializer.js [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+    resourceHash.relationships[descriptor.key] = {
+      data: {
+        type: descriptor.type,
+        id: context.extractId(modelClass, resourceHash)
+      }
+    };
+  },
+
+  _copyAttributesFromPartialToParent: function(serializedHash, partialHash) {
+    Object.keys(partialHash.data.attributes).forEach(attr => {
```
Comment (igorT) [inline addon/mixins/jsonapi-serializer.js]: Object.assign? Not 100% sure about polyfill state

File: README.md (hunk 12 above)
Comment (Azdaroth) [inline README.md]: Hmm, good idea, could be a thing in next release.

File: addon/mixins/jsonapi-serializer.js (hunk 14 above)
Comment (igorT) [inline addon/mixins/jsonapi-serializer.js]: Or Ember.merge, same as below

File: README.md [hunk 15]
```diff
@@ -46,6 +49,97 @@ In your template you can do
...
+  _normalizePartialRelationship: function(modelClass, resourceHash, descriptor) {
+    // add this function to handle normalization of partial extension as a relationship. Required.
+  },
+
+  _copyAttributesFromPartialToParent: function(serializedHash, partialHash) {
+    // add this function to handle copying attributes from partial extension to top-level hash. Required.
+  },
+
+  _modifyHashBasedOnDescriptorKey: function(serializedHash, descriptorKey) {
+    // add this function to perform optional modification of the serialized hash based on descriptor key. Not required.
```
Comment (igorT) [inline README.md]: Example here would be very helpful

File: addon/mixins/rest-serializer.js [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+  _normalizePartialRelationship: function(modelClass, resourceHash, descriptor) {
+    resourceHash[descriptor.key] = resourceHash.id;
+  },
+
+  _copyAttributesFromPartialToParent: function(serializedHash, partialHash) {
+    Ember.merge(serializedHash, partialHash);
+  },
+
+  _modifyHashBasedOnDescriptorKey: function(serializedHash, descriptorKey) {
+    delete serializedHash[descriptorKey];
```
Comment (igorT) [inline addon/mixins/rest-serializer.js]: Why is this needed?

File: addon/utils/serializer-partial-model-handler.js [hunk 17]
```diff
@@ -0,0 +1,33 @@
+export default {
+  preNormalize: function(modelClass, resourceHash, context, normalizePartialRelationshipStrategy) {
+    if (modelClass._isPartialModel) {
```
Comment (igorT) [inline addon/utils/serializer-partial-model-handler.js]: Maybe move this up, so we don't enter this code path if no partial model handling

File: addon/utils/serializer-partial-model-handler.js [hunk 18]
```diff
@@ -0,0 +1,33 @@
+export default {
+  preNormalize: function(modelClass, resourceHash, context, normalizePartialRelationshipStrategy) {
```
Comment (igorT) [inline addon/utils/serializer-partial-model-handler.js]: I would call this something like normalizeEachPartialRelationship which is it's role, rather then by position which is before normalize. When you see the code, you can tell it's called before normalize, so the name doesn't really help you

File: addon/utils/serializer-partial-model-handler.js [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+    if (modelClass._isPartialModel) {
+      partialDescriptors(modelClass).forEach(descriptor => {
+        normalizePartialRelationshipStrategy(modelClass, resourceHash, descriptor, context);
+      });
+    }
+  },
+
+  postSerialize: function(snapshot, serializedHash, copyAttributesFromPartialToParentStrategy,
```
Comment (igorT) [inline addon/utils/serializer-partial-model-handler.js]: Same here re: naming

File: addon/utils/serializer-partial-model-handler.js [hunk 20]
```diff
@@ -0,0 +1,33 @@
+export default {
+  preNormalize: function(modelClass, resourceHash, context, normalizePartialRelationshipStrategy) {
+    if (modelClass._isPartialModel) {
+      partialDescriptors(modelClass).forEach(descriptor => {
+        normalizePartialRelationshipStrategy(modelClass, resourceHash, descriptor, context);
```
Comment (igorT) [inline addon/utils/serializer-partial-model-handler.js]: You need to `.call` `normalizePartialRelationshipStrategy` otherwise if someone is accessing `this` in their callback, it will explode

File: addon/utils/serializer-partial-model-handler.js [hunk 21]
```diff
@@ -0,0 +1,33 @@
...
+      });
+    }
+  },
+
+  postSerialize: function(snapshot, serializedHash, copyAttributesFromPartialToParentStrategy,
+                          modifyHashBasedOnDescriptorKeyStrategy) {
+    partialDescriptors(snapshot).forEach(descriptor => {
+      let partial = snapshot.belongsTo(descriptor.key);
+
+      modifyHashBasedOnDescriptorKeyStrategy(serializedHash, descriptor.key);
```
Comment (igorT) [inline addon/utils/serializer-partial-model-handler.js]: Same here re: `.call`

File: addon/utils/serializer-partial-model-handler.js [hunk 22]
```diff
@@ -0,0 +1,33 @@
...
+      if (partial) {
+        let partialHash = partial.record.serialize();
+        copyAttributesFromPartialToParentStrategy(serializedHash, partialHash);
+      }
+    });
+  }
+};
+
+function partialDescriptors(modelClass) {
+  if (modelClass._partialDescriptors) {
```
Comment (igorT) [inline addon/utils/serializer-partial-model-handler.js:21]: Why do we need this?

File: addon/mixins/rest-serializer.js (hunk 16 above)
Comment (Azdaroth) [inline addon/mixins/rest-serializer.js]: Hmm, actually it isn't, all tests pass, it was present before refactoring, so I moved it here, but in that case the entire function is not needed, good call.

File: addon/mixins/base-serializer.js [hunk 23]
```diff
@@ -0,0 +1,31 @@
...
+
+  serialize: function(snapshot /*, options */) {
+    let serializedHash = this._super(...arguments);
+    postSerialize(snapshot, serializedHash, this._copyAttributesFromPartialToParent,
+      this._modifyHashBasedOnDescriptorKey);
+
+    return serializedHash;
+  },
+
+  serializeBelongsTo: function(/*snapshot, json, relationship */) {
```
Comment (igorT) [inline addon/mixins/base-serializer.js]: You will break everyone's relationship serializing by doing this. Need to comment out.

File: addon/mixins/rest-serializer.js (hunk 16 above)
Comment (igorT) [inline addon/mixins/rest-serializer.js]: I think it passes because of the faulty empty serializeBelongsTo in the base

Comment (igorT) [inline addon/mixins/rest-serializer.js]: I think it should probably be
`delete serializedHash[this.keyForRelationship(descriptorKey, 'belongsTo')]` instead of 
`delete serializedHash[descriptorKey];`

**Comment by igorT** [comment]
I like the structure :+1:

File: addon/utils/serializer-partial-model-handler.js (hunk 22 above)
Comment (Azdaroth) [inline addon/utils/serializer-partial-model-handler.js:21]: We need to extract partial descriptors either from modelClass or snapshot, so that's helper function for some unified way of accessing it and returning empty array if there are no partial descriptors so that it doesn't break in any case. Maybe the argument name could be `modelClassOrSnapshot` so it's closer to what it actually might be.

File: addon/mixins/rest-serializer.js (hunk 16 above)
Comment (Azdaroth) [inline addon/mixins/rest-serializer.js]: Yeah, it does because of the `serializeBelongsTo`, great catch.

Comment (Azdaroth) [inline addon/mixins/rest-serializer.js]: This function seems not to be needed any longer as we could simply handle it in base serializer:

```
serializeBelongsTo: function(snapshot, json, relationship) {
    if (relationship.options.isPartialExtension) {
      return null;
    } else {
      return this._super(...arguments);
    }
  },
```

and even remove from jsonapi-serializer / rest-serializer as already handled in the base one, one function less for customizing other serializers.

Comment (igorT) [inline addon/mixins/rest-serializer.js]: :+1:

**Comment by Azdaroth** [comment]
Fixed all the things. Thanks @igorT for the review.

**Comment by igorT** [comment]
:+1:

**Review by sara554-beep — APPROVED** [review]

**Review by sara554-beep — COMMENTED** [review]
Al no saber realizar cambios directos revisare
